sparrow_datums/keypoints/keypoints.py

```diff
@@ -0,0 +1,191 @@
+        if self.is_absolute:
+            return self
+        x = self.array.copy().astype(np.float32)
+        x[..., :2] *= self.scale
+        if self.ptype is PType.unknown:
+            param_type = PType.absolute_xy
+        else:
+            param_type = self.ptype.as_absolute
+        return self.__class__(
+            x,
+            ptype=param_type,
+            **self.metadata_kwargs,
+        )
+
+    def validate_known_ptype(self) -> None:
+        """Make sure PType is a known box parameterization."""
+        known_box_parameterizations = {PType.absolute_xy, PType.relative_xy}
+        if self.ptype not in known_box_parameterizations:
+            raise ValidationError(f"Unknown box parameterization: {self.ptype.name}")
+
+    @property
+    def x(self) -> FloatArray:
+        """Slice the x dimension of the keypoint."""
+        self.validate_known_ptype()
+        result: FloatArray
+        result = self.array[..., 0]
+        return result
+
+    @property
+    def y(self) -> FloatArray:
+        """Slice the y dimension of the keypoint."""
+        self.validate_known_ptype()
+        result: FloatArray
+        result = self.array[..., 1]
+        return result
+
+    def generate_heatmap(self, x0: int, y0: int, covariance: float):
+        """Create a 2D heatmap from an x, y pixel location.
+
+        Note: This is a helper function for to_heatmap_array()
+        Parameters
+        ----------
+        x0 :
+            x coordinate of the keypoint to be transformed
+        y0 :
+            y coordinate of the keypoint to be transformed
+        covariance :
+            covariance of the surface to be created
+        Returns
+        -------
+            keypoint in form of a 2D array( a heatmap)
+        """
+        img_w = self.image_width
+        img_h = self.image_height
+        xx, yy = np.meshgrid(np.arange(img_w), np.arange(img_h))
+        zz = (
+            1
+            / (2 * np.pi * covariance**2)
+            * np.exp(
+                -(
+                    (xx - x0) ** 2 / (2 * covariance**2)
+                    + (yy - y0) ** 2 / (2 * covariance**2)
+                )
+            )
+        )
+        # Normalize zz to be in [0, 1]
+        zz_min = zz.min()
+        zz_max = zz.max()
+        zz_range = zz_max - zz_min
+        if zz_range == 0:
+            zz_range += 1e-8
+        return (zz - zz_min) / zz_range
+
+    def to_heatmap_array(self, covariance: float = 20) -> npt.NDArray[np.float64]:
+        """Convert Keypoints into a heatmap array.
+
+        Parameters
+        ----------
+        covariance :
+            Covariance of the surface to be created, by default 20
+        Returns
+        -------
+            heatmaps in numpy form.
+        """
+        xs = self.array[..., 0]
+        ys = self.array[..., 1]
+        if np.size(xs) > 1:
+            heatmaps = []
+            for x0, y0 in zip(xs, ys):
```

Comment:
I think this will fail in unexpected ways if the keypoint chunk is 3D or higher. Also, what happens with a single Keypoint e.g:

```
Keypoint(np.ones(2), PType.absolute_xy, image_width=10, image_height=10).to_heatmap_array()
```

Ideally, we would handle those cases in this method. If that's not feasible, we should throw a helpful error message so users know why this method doesn't work for them.

Reply:
- Single Keypoint has been tested [here](https://github.com/ktjayamanna/sparrow-datums/blob/OS-526V2/sparrow_datums/keypoints/keypoints_test.py#L87)
- I couldn't think of a scenario where keypoint chunk goes over 2D. Do you have any examples in mind we could address? If not, we could just raise an exception. Thoughts?

Reply:
Ok nice. Yeah the example would be if there is a batch of n keypoints per image or keypoint tracking. It's fine to not support these for now, but we should raise a helpful exception. I think I would actually put that in the `validate()` function. Something like:

```
if self.ndim > 2:
    raise ValidationError("2D Keypoint chunks are the maximum dimension supported.")
```

Reply:
@jbencook  implemented the exception for >2D dim not supported. Let me know what else you need.